portfolio/src/main/java/com/google/sps/servlets/ChatServlet.java

```diff
@@ -0,0 +1,101 @@
+// Copyright 2019 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.sps.servlets;
+
+import com.google.sps.data.Message;
+import com.google.sps.data.Chat;
+import com.google.gson.Gson;
+import com.google.gson.GsonBuilder;
+import java.io.IOException;
+import java.util.Map;
+import java.util.HashMap;
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+/** Servlet that encapsulates the subtraction game. */
+@WebServlet("/send-message")
+public final class ChatServlet extends HttpServlet {
+  static final String ANONYM = "anonym";
+
+  private Chat chat = new Chat();
+  private String warning = "";
+
+  @Override
+  public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    response.setContentType("application/json");
+    
+    Map<String, Object> data = new HashMap<String, Object>();
+    data.put("chat", chat);
+    data.put("warning", warning);
+    
+    String json = new Gson().toJson(data);
+    response.getWriter().println(json);
+  }
+
+  @Override
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    // Get the input from the form.
+    Message message = getMessage(request);
+
+    warning = "";
+    if (message.nameIsEmpty()) {
+      warning = "Please choose name other than an anonym.\n";
+    }
+    if (message.textIsEmpty()) {
+      warning += "Please write a message.\n";
+    }
+
+    if (warning.length() != 0) {
+      response.sendRedirect("/chat.html");
+      return;
+    }
```

Comment:
Better use HTTP return codes in case of errors. Use either sendError(status, msg) or setStatus(status) from the response object: https://docs.oracle.com/javaee/7/api/javax/servlet/http/HttpServletResponse.html

Have a look at the available status code: https://www.restapitutorial.com/httpstatuscodes.html
You probably want to use one of the 4xx codes which represents errors from the client, e.g. 400 Bad Request with an error message that tells the client that name or text is empty.

Reply:
Hm, I don't want to raise an error. I just don’t want to send this message and print warning for user near the form. Is it wrong approach?

Reply:
Is it possible to set the status code, e.g. to 422, and still print the warning for the user?

Here is a nice related writeup: https://stackoverflow.com/a/59527615